File: 74. Search a 2D Matrix/74. Search a 2D Matrix.dart
class Solution {
  bool searchMatrix(List<List<int>> matrix, int target) {
    if (matrix.isEmpty || matrix[0].isEmpty) {
      return false;
    }
    
    int m = matrix.length;
    int n = matrix[0].length;
    int left = 0;
    int right = m * n - 1;
    
    while (left <= right) {
      int mid = left + (right - left) ~/ 2;
      int midValue = matrix[mid ~/ n][mid % n];
      
      if (midValue == target) {
        return true;
      } else if (midValue < target) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }
    
    return false;
  }
}

File: 74. Search a 2D Matrix/74. Search a 2D Matrix.rs
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        if matrix.is_empty() || matrix[0].is_empty() {
            return false;
        }
        
        let m = matrix.len();
        let n = matrix[0].len();
        let mut left: usize = 0;
        let mut right: usize = m.checked_mul(n).map_or(0, |v| v.saturating_sub(1));
        
        while left <= right {
            let mid = left + (right - left) / 2;
            let mid_value = matrix[mid / n][mid % n];
            
            if mid_value == target {
                return true;
            } else if mid_value < target {
                left = mid.saturating_add(1);
            } else {
                if mid == 0 {
                    break;
                }
                right = mid - 1;
            }
        }
        
        false
    }
}